pkg/collector/process.go

```diff
@@ -361,47 +374,95 @@ func (cp *CollectingProcess) decodeDataSet(dataBuffer *bytes.Buffer, obsDomainID
+	if _, ok := cp.templatesMap[obsDomainID]; !ok {
+		cp.templatesMap[obsDomainID] = make(map[uint16]*template)
 	}
 	elements := make([]*entities.InfoElement, 0)
 	for _, elementWithValue := range elementsWithValue {
 		elements = append(elements, elementWithValue.GetInfoElement())
 	}
-	cp.templatesMap[obsDomainID][templateID] = elements
-	// template lifetime management
+	tpl, ok := cp.templatesMap[obsDomainID][templateID]
+	if !ok {
+		tpl = &template{}
+		cp.templatesMap[obsDomainID][templateID] = tpl
+	}
+	tpl.ies = elements
+	klog.V(4).InfoS("Added template to template map", "obsDomainID", obsDomainID, "templateID", templateID)
+	// Template lifetime management for UDP.
 	if cp.protocol != "udp" {
 		return
 	}
-	// Handle udp template expiration
-	go func() {
-		ticker := time.NewTicker(cp.templateTTL)
-		defer ticker.Stop()
-		select {
-		case <-ticker.C:
+	tpl.expiryTime = cp.clock.Now().Add(cp.templateTTL)
+	if tpl.expiryTimer == nil {
+		tpl.expiryTimer = cp.clock.AfterFunc(cp.templateTTL, func() {
 			klog.Infof("Template with id %d, and obsDomainID %d is expired.", templateID, obsDomainID)
-			cp.deleteTemplate(obsDomainID, templateID)
-			break
+			now := cp.clock.Now()
+			// From the Go documentation:
+			//   For a func-based timer created with AfterFunc(d, f), Reset either
+			//   reschedules when f will run, in which case Reset returns true, or
+			//   schedules f to run again, in which case it returns false. When Reset
+			//   returns false, Reset neither waits for the prior f to complete before
+			//   returning nor does it guarantee that the subsequent goroutine running f
+			//   does not run concurrently with the prior one. If the caller needs to
+			//   know whether the prior execution of f is completed, it must coordinate
+			//   with f explicitly.
+			// In our case, when f executes, we have to verify that the record is indeed
+			// scheduled for deletion by checking expiryTime. We cannot just
+			// automatically delete the template.
+			cp.deleteTemplateWithConds(obsDomainID, templateID, func(tpl *template) bool {
+				// lock will be held when this executes
+				return !tpl.expiryTime.After(now)
+			})
+		})
+	} else {
+		tpl.expiryTimer.Reset(cp.templateTTL)
+	}
+}
+
+// deleteTemplate returns true iff a template was actually deleted.
+func (cp *CollectingProcess) deleteTemplate(obsDomainID uint32, templateID uint16) bool {
+	return cp.deleteTemplateWithConds(obsDomainID, templateID)
+}
+
+// deleteTemplateWithConds returns true iff a template was actually deleted.
+func (cp *CollectingProcess) deleteTemplateWithConds(obsDomainID uint32, templateID uint16, condFns ...func(*template) bool) bool {
+	cp.mutex.Lock()
+	defer cp.mutex.Unlock()
+	template, ok := cp.templatesMap[obsDomainID][templateID]
+	if !ok {
+		return false
+	}
+	for _, condFn := range condFns {
+		if !condFn(template) {
+			return false
 		}
-	}()
+	}
+	// expiryTimer will be nil when the protocol is UDP.
```

Comment:
Could you remind me when will the expiryTimer be set to nil when the protocol is UDP?

Reply:
It's a typo, it should be TCP

Reply:
Fixed, thanks for catching it